Math/NumberTheory/Recurrencies/Pentagonal.hs

```diff
@@ -0,0 +1,74 @@
+-- Values of <https://en.wikipedia.org/wiki/Partition_(number_theory)#Partition_function partition function>.
+
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Math.NumberTheory.Recurrencies.Pentagonal
+  ( partition
+  , pentagonalSigns
+  , pents
+  ) where
+
+import qualified Data.Map as M
+import Numeric.Natural (Natural)
+
+-- | Infinite list of generalized pentagonal numbers.
+-- Example:
+--
+-- >>> take 10 pents
+-- [0, 1, 2, 5, 7, 12 ,15, 22, 26, 35]
+pents :: (Enum a, Num a) => [a]
+pents = interleave (scanl (\acc n -> acc + 3 * n - 1) 0 [1..])
+                   (scanl (\acc n -> acc + 3 * n - 2) 1 [2..])
+  where
+    interleave :: [a] -> [a] -> [a]
+    interleave (n : ns) (m : ms) = n : m : interleave ns ms
+    interleave _ _ = []
+
+-- | When calculating the @n@-th partition number @p(n)@ using the sum
+-- @p(n) = p(n-1) + p(n-2) - p(n-5) - p(n-7) + p(n-11) + ...@, the signs of each
+-- term alternate every two terms, starting with a positive sign.
+-- @pentagonalSigns@ takes a list of numbers and produces such an alternated
+-- sequence.
+-- Examples:
+--
+-- >>> pentagonalSigns [1..5]
+-- [1, 2, -3, -4, 5]
+--
+-- >>> pentagonalSigns [1..6]
+-- [1, 2, -3, -4, 5, 6]
+pentagonalSigns :: Num a => [a] -> [a]
+pentagonalSigns = zipWith (*) (cycle [1, 1, -1, -1])
+
+-- @partition !! n@ calculates the @n@-th partition number:
+-- @p(n) = p(n-1) + p(n-2) - p(n-5) - p(n-7) + p(n-11) + ...@, where @p(0) = 1@
+-- and @p(k) = 0@ for a negative integer @k@. Uses a @Map@ from the
+-- @containers@ package to memoize previous results.
+-- 
+-- Example: calculating @partition !! 10@, assuming the memoization map is
+-- filled and called @dict :: Integral a => Map a a@.
+-- 
+-- * @tail [0, 1, 2, 5, 7, 12 ,15, 22, 26, 35, ..] == [1, 2, 5, 7, 12 ,15, 22, 26, 35, 40, ..]@.
+-- * @takeWhile (\m -> 10 - m >= 0) == [1, 2, 5, 7]@.
+-- * @map (\m -> dict ! fromIntegral (10 - m)) [1, 2, 5, 7] == [dict ! 9, dict ! 8, dict ! 5, dict ! 3] == [30, 22, 7, 3]@
+-- * @pentagonalSigns [30, 22, 7, 3] == [30, 22, 7, 3] == [30, 22, -7, -3]@
+-- * @sum [30, 22, -7, -3] == 42@
+--
+-- Note: @tail@ is applied to @pents@ because otherwise the calculation of
+-- @p(n)@ would involve a duplicated @p(n-1)@ term (see the above example).
+partition :: forall a . (Enum a, Num a, Ord a) => [a]
+partition = 1 : go (M.insert 0 1 mempty) 1
+  where
+    go :: (Enum a, Num a, Ord a) => M.Map a a -> a -> [a]
+    go dict !n =
+        let n' = (sum .
+                 pentagonalSigns .
+                 map (\m -> dict M.! (n - m)) .
+                 takeWhile (\m -> n - m >= 0) .
```

Comment:
> @Bodigrim can you try doing take 10 (Math.NumberTheory.Recurrencies.partition) :: [Math.NumberTheory.Moduli.Mod 10] in a cabal new-repl session?
>
> It calculates the head of the list, and that's as far as it goes. Consumed 12GB of RAM before I killed the session.

I can reproduce the issue. AFAIU it happens because the condition `n - m >= 0 :: [Mod m]` is always true in modular arithmetic. Can we actually employ `Map Int a` or even `IntMap a` for caching here?

Reply:
>Can we actually employ Map Int a or even IntMap a for caching here?
Are you asking if it is feasible, or if it is at all what we want to do?

Reply:
My impression is that it is possible and desirable. `maxBound :: Int` is big enough to store indices, so no worries about possible overflow here.

Reply:
Even after using `Map Int a`, doing `partition :: [Mod 10]` won't calculate more than 7 terms. I think it's because of the following line, `tail) pents`. By asking for `partition` with type `[Mod 10]`, what happens is that `pents` also has that type, and this spoils the remainder of the calculation. The following `takeWhile` won't stop where it's supposed to, and so on.

I'll try giving `pents` type `[Int]` and see what happens.